### PR TITLE
[WIP] Fix issue #384 (invisible legends when y is negative)

### PR DESCRIPTION
**Pull request by n-riesco**
TODO

* [ ] Rename `Plot.autoMarginVertical()` to something more sensible (any suggestions?)
* [x] Add missing `select('rect')`
* [x] Debug test failure: `pie hovering event data should contain the correct fields FAILED`
* [x] Debug test failure: `pie hovering event data should fire when moving from one slice to another FAILED`
* [x] Debug test failure: `the range slider when specified as visible should react to resizing the maximum handle FAILED`
* [ ] Check image test failures  (currently 27, 5, autorange-tozero-rangemode, and range-selector-style. All these mocks locate the legend above the axis, i.e. y > 1).
* [ ] New tests

### CODE REVIEW

File: src/components/legend/draw.js [hunk 1]
```diff
@@ -184,27 +184,42 @@ module.exports = function draw(gd) {
...
+    lx = Math.round(lx);
+    ly = Math.round(ly);
+
+    // Make sure the legend top is below the top margin
+    if(ly < fullLayout.margin.t) ly = fullLayout.margin.t;
+
+    var scrollHeightMax = fullLayout.height - fullLayout.margin.b - ly;
+    var scrollHeight = Math.min(scrollHeightMax, opts.height);
+
+    if(scrollHeight <= 2 * opts.borderwidth) {
```
Comment (n-riesco) [inline src/components/legend/draw.js]: On second thoughts, it'd be more user-friendly to provide a safe default here. For example:
- if `opts.height >= fullLayout.height`, then `ly = fullLayout.margin.t; scrollHeight = fullLayout.height;`
- else locate the legend the bottom.

File: src/components/legend/draw.js [hunk 2]
```diff
@@ -184,27 +184,42 @@ module.exports = function draw(gd) {
...
+        console.error('Legend.draw: insufficient space to draw legend');
+        legend.remove();
+        return;
+    }
```
Comment (n-riesco) [inline src/components/legend/draw.js]: On second thoughts, it'd be more user-friendly to provide a safe default here. For example:
- if `opts.height >= fullLayout.height`, then `ly = fullLayout.margin.t; scrollHeight = fullLayout.height;`
- else locate the legend the bottom.

**Comment by n-riesco** [comment]
The location of the plot title seems to be affected by the same issue that here I'm fixing for legends.

I've checked and 3 out of the 4 failing mocks (5, 27 and autorange-tozero-rangemode) have a title that overlaps the top margin.

I haven't checked yet, but I suspect that the 4th failing mock (range-selector-style) suffers from a similar problem with the bottom margin.

File: src/plots/plots.js [hunk 3]
```diff
@@ -875,6 +875,36 @@ plots.autoMargin = function(gd, id, o) {
     }
 };
 
+// Similar to plots.autoMargin, except that it pads the request with the layout
+// margins, so that an object can be drawn without reaching the layout vertical
+// margins.
+plots.autoMarginVertical = function(gd, id, o) {
```
Comment (etpinard) [inline src/plots/plots.js]: why do we need this?

**Comment by etpinard** [comment]
I'm not convinced we should add a different auto-margin step to accommodate negative legend positions.

As fas as I know, negative legend positions were working fine before https://github.com/plotly/plotly.js/pull/243 (or in https://github.com/plotly/plotly.js/pull/365). It's really a shame that we don't have any image test for those cases. 

But, it looking back at `components/legend/draw.js` before https://github.com/plotly/plotly.js/pull/243 should tells us where issue #384 was introduced.

**Comment by n-riesco** [comment]
@etpinard I've checked v1.5.2 (#243 was introduced after) and the issue is not present.

I have a question though. How strict are the margins defined in `layout.margin`? Are components, like legends and titles, allowed to stretch into the margins. If they are not allowed to overlap the margins, then there is an issue in `Plots.autoMargin`. This is the reason why I introduced the ill-named `Plots.autoMarginVertical`.

Shall I remove `Plots.autoMarginVertical`?

**Comment by etpinard** [comment]
@n-riesco 

> I've checked v1.5.2 (#243 was introduced after) and the issue is not presen

Thanks for the investigation :beers: 

> How strict are the margins defined in layout.margin

No very strict. They need to be [greater than zero](https://github.com/plotly/plotly.js/blob/f692714952568346b6a92e0ff59dc8458cdfe5ac/src/plots/layout_attributes.js#L75-L103) and [not bigger than the plot width and/or height](https://github.com/plotly/plotly.js/blob/f692714952568346b6a92e0ff59dc8458cdfe5ac/src/plots/plots.js#L818-L844)

> Are components, like legends and titles, allowed to stretch into the margins 

It's more like the other around: *components, like legends and titles expand the margins (by default)*.

> If they [components] are not allowed to overlap the margins, then there is an issue in Plots.autoMargin

Components **are allow** to overlap the margins if that's what the users sent as `layout` argument to `Plotly.plot`. That said, plotly.js should attempt to make the margins big enough to fit the component by default (as done currently in [`Plots.autoMargin`](https://github.com/plotly/plotly.js/blob/f692714952568346b6a92e0ff59dc8458cdfe5ac/src/plots/plots.js#L851)).

> Shall I remove Plots.autoMarginVertical ?

Yes. You should try to fix this issue from within the `legend/draw.js`.

**Comment by etpinard** [comment]
@n-riesco 

> I've checked v1.5.2 (#243 was introduced after) and the issue is not presen

Thanks for the investigation :beers: 

> How strict are the margins defined in layout.margin

No very strict. They need to be [greater than zero](https://github.com/plotly/plotly.js/blob/f692714952568346b6a92e0ff59dc8458cdfe5ac/src/plots/layout_attributes.js#L75-L103) and [not bigger than the plot width and/or height](https://github.com/plotly/plotly.js/blob/f692714952568346b6a92e0ff59dc8458cdfe5ac/src/plots/plots.js#L818-L844)

> Are components, like legends and titles, allowed to stretch into the margins 

It's more like the other around: *components, like legends and titles expand the margins (by default)*.

> If they [components] are not allowed to overlap the margins, then there is an issue in Plots.autoMargin

Components **are allow** to overlap the margins if that's what the users sent as `layout` argument to `Plotly.plot`. That said, plotly.js should attempt to make the margins big enough to fit the component by default (as done currently in [`Plots.autoMargin`](https://github.com/plotly/plotly.js/blob/f692714952568346b6a92e0ff59dc8458cdfe5ac/src/plots/plots.js#L851)).

> Shall I remove Plots.autoMarginVertical ?

Yes. You should try to fix this issue from within the `legend/draw.js`.

**Comment by n-riesco** [comment]
Moved to #417

**Pull request closed without merging**